xapian-core/include/xapian/weight.h

```diff
@@ -573,21 +573,122 @@ class XAPIAN_VISIBILITY_DEFAULT BM25Weight : public Weight {
-			Xapian::termcount uniqterms) const;
-    double get_maxextra() const;
+	double get_sumextra(Xapian::termcount doclen,
+		Xapian::termcount uniqterms) const;
+	double get_maxextra() const;
+
+  protected:
+	// BM25+ is disabled.
+	bool bm25_plus = 0;
+};
+
+/// Xapian::Weight subclass implementing the BM25+ probabilistic formula.
+class XAPIAN_VISIBILITY_DEFAULT BM25PlusWeight : public BM25Weight {
+
+  public:
+	/** Construct a BM25PlusWeight.
+	 *
+	 *  @param delta  A parameter for pseudo tf value to control the scale
+	 *		 of the tf lower bound. Delta(δ) can be tuned for e.g from 0.0
+	 *		 to 1.5 but BM25+ can still work effectively across collections
+	 *		 with a fixed δ = 1.0. (default 1.0)
+	 */
+	BM25PlusWeight(double k1, double k2, double k3, double b, double min_normlen, double delta)
+	: param_k1(k1), param_k2(k2), param_k3(k3), param_b(b),
+	param_min_normlen(min_normlen), param_delta(delta)
+	{
+	if (param_k1 < 0) param_k1 = 0;
+	if (param_k2 < 0) param_k2 = 0;
+	if (param_k3 < 0) param_k3 = 0;
+	if (param_b < 0) {
+	param_b = 0;
+	} else if (param_b > 1) {
+	param_b = 1;
+	}
+	need_stat(COLLECTION_SIZE);
+	need_stat(RSET_SIZE);
+	need_stat(TERMFREQ);
+	need_stat(RELTERMFREQ);
+	need_stat(WDF);
+	need_stat(WDF_MAX);
+	if (param_k2 != 0 || (param_k1 != 0 && param_b != 0)) {
+	need_stat(DOC_LENGTH_MIN);
+	need_stat(AVERAGE_LENGTH);
+	}
+	if (param_k1 != 0 && param_b != 0) need_stat(DOC_LENGTH);
+	if (param_k2 != 0) need_stat(QUERY_LENGTH);
+	if (param_k3 != 0) need_stat(WQF);
+	if (param_delta < 0) param_delta = 0;
+	if (param_delta != 0) {
+	need_stat(AVERAGE_LENGTH);
+	need_stat(DOC_LENGTH);
+	need_stat(WQF);
+	}
+	}
+
+	BM25PlusWeight()
+	: param_k1(1), param_k2(0), param_k3(1), param_b(0.5),
+	  param_min_normlen(0.5), param_delta(1)
+	{
+	need_stat(COLLECTION_SIZE);
+	need_stat(RSET_SIZE);
+	need_stat(TERMFREQ);
+	need_stat(RELTERMFREQ);
+	need_stat(WDF);
+	need_stat(WDF_MAX);
+	need_stat(DOC_LENGTH_MIN);
+	need_stat(AVERAGE_LENGTH);
+	need_stat(DOC_LENGTH);
+	need_stat(WQF);
+	}
+	std::string name() const;
+
+	std::string serialise() const;
+	BM25PlusWeight * unserialise(const std::string & serialised) const;
+
+	double get_sumpart(Xapian::termcount wdf,
+		   Xapian::termcount doclen,
+		   Xapian::termcount uniqterm) const;
+	double get_maxpart() const;
+
+	double get_sumextra(Xapian::termcount doclen,
+		Xapian::termcount uniqterms) const;
+	double get_maxextra() const;
+
+  protected:
+	/// Factor to multiply the document length by.
+	mutable Xapian::doclength len_factor;
+
+	/// Factor combining all the document independent factors.
+	mutable double termweight;
+
+	/// The BM25 parameters.
+	double param_k1, param_k2, param_k3, param_b;
+
+	/// The minimum normalised document length value.
+	Xapian::doclength param_min_normlen;
+
+	/// Additional parameter delta in the BM25+ formula.
+	double param_delta;
+
+	BM25PlusWeight * clone() const;
+
+	void init(double factor);
+
+	// BM25+ is enabled.
+	bool bm25_plus = 1;
```

Comment:
How is this flag useful?

If we are just adding BM25+ support as an option to the existing `BM25Weight` class, then adding a flag to that class would make sense.  But subclassing is an alternative to that.  Doing both isn't useful - methods of `BM25PlusWeight` know they're methods of `BM25PlusWeight`, and methods of the parent class called on a `BM25PlusWeight` will see the member variable of the parent class, not the subclass.

I still feel making this an option in `BM25Weight` is the best way to go, but if we're going for a subclass, can we get rid of this flag?